_posts/2020-01-30-Age-Targeting-Part1.md

```diff
@@ -0,0 +1,185 @@
+However, it is clear that some of the distributions in the plot do not make sense.
+Why would people from 17-18 correspond to more than 95% of the total users of the app? (What happens to the 16ties? or 19ties?)
+Why would apps have the same numbers of devices for every age in a specific range?
+These must be very special apps, or the data do not correspond to truth.
+What we want to do is to exclude these apps.
+A rejection/accept statement is drown only when we have enough users observed in the app.
+In case the number of users is too low we do not consider them in the study and in the final ground truth selection.  
+This is not a big issue since the devices contained in these apps are few by definition, so we do not cut off too many data in this way.
+
+Let's describe the main pattern we discover in the data and how we get rid of them creating ad hoc variable via pyspark functions (spark allows to scale and run over the entire dataset).
+
+
+* **Few ages**<br/>
+Certain apps provide almost always only two/three values.
+We create a variable (frac12) which is the fraction of devices contained in the two top most frequent ages observed in the app.
+Apps where the frac12 value is higher than 30% are excluded.
+Below you can see the distribution of the frac12 variable and the consequences of the rejection of the non reliable apps on the total age distribution.
+![frac12_cut](../images/Age-Targeting/frac12_cut.png)
+The red vertical line indicates the threshold we apply over the frac12 feature.
+Here the code we use to calculate the frac12 over all apps:
+
+```python
+from pyspark.sql import Window
+from pyspark.sql import functions as F
+
+def get_site_with_frac_top_two_bin_information(df_app_dev_age):
+    """Return the apps with the frac12 column which indicate the ammount of deviceid 
+    observed in the two top age."""
+    df_app_age = df_app_dev_age.groupby(["apps", "age"]).agg(
+        F.count("deviceid").alias("num_dev_per_age"))
+    
+    window_num_dev = Window.partitionBy(df_app_age['apps'])
+    df_app_age = df_app_age.withColumn(
+        "num_dev", F.sum("num_dev_per_age").over(window_num_dev))
+    df_app_age = df_app_age.withColumn(
+        "frac_num_age", df_app_age.num_dev_per_age / df_app_age.num_dev)
+    
+    window_frac = Window.partitionBy(df_app_age['apps']).orderBy(
+        df_app_age['frac_num_age'].desc(), "rand")
+    df_app_age = df_app_age.withColumn("rank_frac", F.rank().over(window_frac))
+    df_app_frac12 = df_app_age.filter(df_app_age.rank_frac < 3).groupby("apps").agg(
+        F.sum("frac_num_age").alias("frac12"))
+    return df_app_frac12
+```
+
+* **Uniform distribution**<br/>
+We found out apps where for a fixed device a different age is provided every time the device accesses the app.
+For every devices which have been observed at least twice in the app, we calculate the max - min age value and, then, we calculate the average of this differences.
```

Comment:
For every device which has been observed...

Reply:
we calculate the range of ages, thus the difference between maximum and minimum age reported. Then, we calculate the average the those values for each app.